File: java/socketd/src/main/java/org/noear/socketd/transport/core/codec/ByteBufferCodecReader.java
package org.noear.socketd.transport.core.codec;

import org.noear.socketd.transport.core.CodecReader;

import java.nio.ByteBuffer;

/**
 * 缓冲读 ByteBuffer 适配
 *
 * @author noear
 * @since 2.0
 */
public class ByteBufferCodecReader implements CodecReader {
    private ByteBuffer buffer;

    public ByteBufferCodecReader(ByteBuffer buffer) {
        this.buffer = buffer;
    }

    /**
     * 获取 byte
     */
    @Override
    public byte getByte() {
        return buffer.get();
    }

    /**
     * 获取一组 byte
     */
    @Override
    public void getBytes(byte[] dst, int offset, int length) {
        buffer.get(dst, offset, length);
    }

    /**
     * 获取 int
     */
    @Override
    public int getInt() {
        return buffer.getInt();
    }

    @Override
    public void skipBytes(int length) {
        buffer.position(buffer.position() + length);
    }

    /**
     * 剩余长度
     */
    @Override
    public int remaining() {
        return buffer.remaining();
    }

    /**
     * 当前位置
     */
    @Override
    public int position() {
        return buffer.position();
    }
}

File: python/socketd/socketd/transport/core/codec/ByteBufferCodecReader.py
from io import BytesIO

from socketd.transport.core.codec.Buffer import Buffer
from socketd.transport.core.Codec import CodecReader


class ByteBufferCodecReader(CodecReader):

    def __init__(self, buffer: Buffer):
        self.__buffer = buffer

    def get_bytes(self) -> bytes:
        return self.__buffer.getvalue()

    def get_int(self) -> int:
        return int.from_bytes(self.__buffer.read1(4), byteorder='little', signed=False)

    def skip_bytes(self, size):
        self.__buffer.seek(self.__buffer.tell() + size)

    def remaining(self):
        return self.__buffer.remaining()

    def position(self):
        return self.__buffer.tell()

    def get_buffer(self) -> BytesIO:
        return self.__buffer

    def close(self):
        self.__buffer.close()


